cpp/src/copying/purge_nonempty_nulls.cu

```diff
@@ -0,0 +1,139 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/copying.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/gather.cuh>
+
+#include <thrust/iterator/counting_iterator.h>
+
+namespace cudf {
+namespace detail {
+
+using cudf::type_id;
+
+namespace {
+
+/// Check if nonempty-null checks can be skipped for a given type.
+bool can_have_nonempty_nulls(cudf::type_id const& type)
+{
+  return type == type_id::STRING || type == type_id::LIST || type == type_id::STRUCT;
+}
+
+/// Check if the (STRING/LIST) column has any null rows with non-zero length.
+bool has_nonempty_null_rows(cudf::column_view const& input, rmm::cuda_stream_view stream)
+{
+  if (not input.has_nulls()) { return false; }  // No nulls => no dirty rows.
+
+  // Cross-reference nullmask and offsets.
+  auto const type         = input.type().id();
+  auto const offsets      = (type == type_id::STRING) ? (strings_column_view{input}).offsets()
+                                                      : (lists_column_view{input}).offsets();
+  auto const d_input      = cudf::column_device_view::create(input);
+  auto const is_dirty_row = [d_input = *d_input, offsets = offsets.begin<size_type>()] __device__(
+                              size_type const& row_idx) {
+    return d_input.is_null_nocheck(row_idx) && (offsets[row_idx] != offsets[row_idx + 1]);
+  };
+
+  auto const row_begin = thrust::make_counting_iterator<cudf::size_type>(0);
+  auto const row_end   = row_begin + input.size();
+  return thrust::any_of(rmm::exec_policy(stream), row_begin, row_end, is_dirty_row);
+}
+
+/// Check if any child columns of the `input` need sanitizing.
+bool children_have_nonempty_nulls(cudf::column_view const& input, rmm::cuda_stream_view stream)
+{
+  return std::any_of(input.child_begin(), input.child_end(), [stream](auto const& child) {
+    return cudf::detail::has_nonempty_nulls(child, stream);
+  });
+}
+
+}  // namespace
+
+/**
+ * @copydoc cudf::detail::has_nonempty_nulls
+ */
+bool has_nonempty_nulls(cudf::column_view const& input, rmm::cuda_stream_view stream)
+{
+  auto const type = input.type().id();
+
+  if (not can_have_nonempty_nulls(type)) { return false; }
+
+  // For types with variable-length rows, check if any rows are "dirty".
+  // A dirty row is a null row with non-zero length.
+  if ((type == type_id::STRING || type == type_id::LIST) && has_nonempty_null_rows(input, stream)) {
+    return true;
+  }
+
+  // For complex types, check if child columns need purging.
+  if ((type == type_id::STRUCT || type == type_id::LIST) &&
+      children_have_nonempty_nulls(input, stream)) {
+    return true;
+  }
```

Comment:
Is there a recursive step I'm missing somewhere? It seems like this will only check the first level of children. What about the further levels of children?

Reply:
Oh, yeah, I did miss the recursive step in `children_have_nonempty_nulls`. tbh, I'd just put that algorithm here instead of moving it to a function.

Reply:
```suggestion
  if ((type == type_id::STRUCT || type == type_id::LIST)  {
    return std::any_of(input.child_begin(), input.child_end(), [stream](auto const& child) {
       return has_nonempty_nulls(child, stream);
  });
  }
```